source/_posts/2016-01-25-upgrading-to-ruby-on-rails-5-0.markdown

```diff
@@ -0,0 +1,226 @@
+```
+
+We execute `bundle update rails` again and now it installs everything correctly.
+
+Now we can run the tests, for example.
+Hence, we execute `bin/rails test` (this is the new way to run tests in Rails 5!) and we have various tests failing.
+
+
+## Problem 1 - the tests fail and we don’t have any further information
+
+From this point forward, the Rails guides give us no further information and we still have various tests failing. In particular, failing is a series of tests of the following form ...
+
+```
+ActionController::RoutingError: No route matches [POST] "/accounts/1/transactions/1"
+```
+
+The next thing we can do is force our application to use [Rails master](https://github.com/rails/rails) in order to see if some of the test failures are due to problems already solved.
+So, we substitute the Gemfile Rails entry with `gem 'rails', github: 'rails/rails'` and once again execute `bundle update rails`.
+
+We return to running the tests with `bin/rails test` and the tests that fail with RoutingError disappear.
+There are, however, other tests that continue failing.
+And now, we have no choice but to begin investigating the errors one by one. 
+
+
+## Problem 2 - expected Array (got String) for param xxx
+
+One of the errors that `bin/rails test` gives us is ...
+
+```
+E
+
+ActionController::BadRequest: Invalid request parameters: expected Array (got String) for param `kind'
+
+bin/rails test test/features/entries/create_entry_test.rb:38
+```
+
+The feature test tries to create an Entry from the form. An Entry in our system is a movement of money that, among other attributes, has a kind attribute with a value of expense, income or transfer. The kind field is an enum in the model and is represented by a radio button in the form. The error indicates that for some reason Rails is expecting an Array when it should be expecting a String.
+
+The first thing that to me seemed reasonable to review was, what is being sent to the server and how. The HTML generated for this form is as follows ...
+
+```
+<input value="income" name="entry[kind]" id="entry_kind" type="radio">
+```
+
+which is correct, but this also appears ...
+
+```
+<input name="entry[kind][]" value="" type="hidden">
+```
+
+The hidden is for forcing an empty value to be sent to the server even when the user isn’t clicking one of the options that the radio buttons present.
+This is fine, but [] appear at the end of the input’s name, which causes the data to arrive to the server as an Array when it should arrive as a String - this is what’s incorrect. 
+
+Researching the changes in Rails a bit, we find ourselves with a regression inserted in [this commit](https://github.com/rails/rails/commit/491013e06d0ad4a296cc23be6c4a48bb0a98106f). This regression isn’t occurring because of the hidden field having been inserted via the radio buttons, but rather it appears because we are reusing all of the code used by the checkboxes, which need to send an Array of elements and not a String.
```

Comment:
I would write this part in passive form, so we don't say "we are reusing all the code.." (since it's Rails' code).

`but rather it appears because all of the code used by the checkboxes is reused, which need to send an Array of elements and not a String.`